### ISSUE TITLE
TMC2209-V1.3

### ISSUE THREAD

**Issue by davcolme** [bigtreetech/BIGTREETECH-TMC2209-V1.2#16]
I have several TMC 2209 V1.3 drivers and I want to connect them via UART via hardwire. TMCs come with 3 pins marked RX, TX, and CLK. Using the RX or TX pins with and without a 1K resistor I am not able to connect to the module via UART. They do not have documentation on the github of said modules. Please, can you post the documentation and indicate how the connection is?

**Comment by Emerica** [comment]
I'd like to second this request. - Boards are Labeled 1.3 - Box is labeled 1.2 X0038B8MA9
I've got an Anycubic Mega S  - Trigorilla 0.0.2 - Marlin 2+ , 2.1.2now
I've tried for a portion of today to get these talking, they appear to function, just no UART comms. Are they addressed different or something?
One of the approaches I've seen is here but it's for the 2208's
https://ohkin.mydns.jp/archives/1072?unapproved=1545&moderation-hash=163a2dc04cda7f6088f3fa20b1a1070b#comment-1545

I'm trying a single stepstick on X, tried hardware and software serial,  with and without resistors to multiple locations, not getting anywhere.

I've noted that RX on the stepstick does in fact reach the UART pin on the chip, but the TX pin seems to go no where?
There's a un-jumped jumper on the board back to RX?

Limited research is telling me the SKR boards move a jumper to a weird position to deal with these. What are they jumping?

**Comment by Emerica** [comment]
I managed to connect one of my drivers up to a stock Arduino Mega 2650 with <TMCStepper.h>
Talking over Serial1    18/19  using the Y cable with 1K to the 4th pin next to MS2.
I have En grounded, 
dir, step, clk, ms1, ms2, diag floating
It's addressed at 0b00

Appears to be working? I can set current levels and stepping, my PSU and the stepper confirm the changes.
This only works while the external power is applied. 

So either I've got Marlin issues, or the Trigorilla board is setting pins on the driver to states that breaks the UART comms, or has issues with Serial connections.

I've noted the EN pin on my board to be HIGH, not sure if it switches LOW to enable the driver and check for status.

**Comment by dylanede** [comment]
I have the same problem. @davcolme did you ever find a solution to this? @Emerica any ideas, considering you found one setup where it worked?

When trying to use the UART I have dir, step, clk all set low, and ms1, ms2 floating. I've tried the single UART wire connected to the TX and the RX pins, with no response from the driver.

The driver chip does not appear to be faulty as pulsing the step pin to move the motor works.

**Comment by davcolme** [comment]
hi @dylanede 

I still can't solve the problem

**Comment by nmaas87** [comment]
Does not work, can confirm

**Comment by Emerica** [comment]
I don't think I ended up getting it all to work with marlin. The drivers did work with stock arduino, so I don't think I can fault BTT.  I was actually fighting two printers so my mind's slightly foggy.
The Anycubic/Trigoriila is the one I mentioned above and I think I said F' it and put them in standalone and then shelved it for now. I've got more pi's now so maybe I'll get klipper loaded and try again there.

The other printer was with a ender 3 that I uart modded, and that had 2209s which I had problems getting all the axis to work at the same time and compile right, I had to use the deprecated maple compile, and eventually just switched to klipper with a switchwire conversion, worked out of the box...

Edit-plugged in the Anycubic and it's TMC Com Error, so yeah, no, didn't end up anywhere.

**Comment by reinux** [comment]
Spent an entire day and half reading about this thing everywhere and ended up memorizing the entire pinout and several arduino tmc2209 controller libs APIs.

Still can't get uart working.

The last thing I have yet to try is the jumper on the back. The instructions in the manual aren't very well translated.

Is It just for using rx tx as opposed to single cable uart?

**Comment by nmaas87** [comment]
Hey @reinux - I only got this working in connecting to the TMC2209 like shown here: https://github.com/janelia-arduino/TMC2209?tab=readme-ov-file#bidirectional-communication with the resistor. I think I connected this to the PDN_RX/Pin 5 on the TMC2209 as the UART "side" shown in the graphic. I tried the TMC2009 lib this graphic comes from, but it was unstable with RP2040 back in the time. What worked however (just to get the feet wet) was Watterotts TMC220x configuration tool together with a usual TTL Level UART USB interface - https://learn.watterott.com/silentstepstick/configurator/ - with that one I could access the UART interface and play around with them.
Later on I decided to use the user manual and write my own UART binary sequences to configure my TMC2209 more reliable via an RP2040 (just pumping out the sequences as needed...). Hope this helps a bit.

**Comment by Emerica** [comment]
I converted the Anycubic to klipper too and that *solved* it for me.

**Comment by reinux** [comment]
> Hey [@reinux](https://github.com/reinux) - I only got this working in connecting to the TMC2209 like shown here: https://github.com/janelia-arduino/TMC2209?tab=readme-ov-file#bidirectional-communication with the resistor. I think I connected this to the PDN_RX/Pin 5 on the TMC2209 as the UART "side" shown in the graphic. I tried the TMC2009 lib this graphic comes from, but it was unstable with RP2040 back in the time. What worked however (just to get the feet wet) was Watterotts TMC220x configuration tool together with a usual TTL Level UART USB interface - https://learn.watterott.com/silentstepstick/configurator/ - with that one I could access the UART interface and play around with them. Later on I decided to use the user manual and write my own UART binary sequences to configure my TMC2209 more reliable via an RP2040 (just pumping out the sequences as needed...). Hope this helps a bit.

Thanks!

I actually _just_ managed to get it working by soldering the two jumper pads on the back together. I was connecting RX/TX with the resistor to Pin 4, which is labeled RX, but that wasn't working.

What I'm wondering now, based on what you're telling me, is if maybe those pins are mislabeled or something and I was supposed to be using Pin 5, and shorting them together actually just means that I'm actually connecting to Pin 5 now.

**Comment by nmaas87** [comment]
Don't worry @reinux - I just got hold of my board again, you're right, I directly connected to Pin 5, labeled "RX" on the silkscreen and did not close a jumper, so - I think pin 4 ("TX") does nothing?

Actually, yes, it does not - because it is probably not connected unless you close the jumper on the back. Because the TMC2209 only has one UART pin ("PDN_UART"):

<img width="612" height="561" alt="Image" src="https://github.com/user-attachments/assets/36bcd6af-ed64-4abb-a2e7-163a51b09c51" />

So, basically: Just connect your MCU like this to your TMC2209, while using the pin called RX as the UART port and you should be good to go.

<img width="270" height="97" alt="Image" src="https://github.com/user-attachments/assets/acbd61ab-f84c-4337-8d06-1c78b0e3ee35" />

I guess the "TX" pin is only there for legacy support of different 3D printer motherboards - but probably needs a 1k resistor on the jumper pad on the back so that you actually get a "seperated" RX/TX pin. But if you're printer motherboard is already having this resistor bridge, because it was purposely made for TMC220x (I guess the 2208 might be the same with one PDN_UART?) - then you can just drop it in and forget.

I hope this makes sense and is finally a solution :)

**Comment by reinux** [comment]
Makes sense. In my case I'm actually using an ESP32, which probably complicated things a little more. Thanks!
